cli/medperf/comms/entity_resources/resources.py

```diff
@@ -1,135 +1,177 @@
+        img_storage = os.path.join(imgs_storage, hash_value)
         shutil.move(image_cube_file, img_storage)
     else:
-        img_storage = os.path.join(imgs_storage, hash)
-
-    if not os.path.exists(img_storage):
-        # If image doesn't exist locally, download it normally
-        # And move it to shared storage
-        _, local_hash = get_cube_file(url, cube_path, image_path, image_name)
-        shutil.move(image_cube_file, img_storage)
+        img_storage = os.path.join(imgs_storage, hash_value)
+        if not os.path.exists(img_storage):
+            # If image doesn't exist locally, download it normally
+            download_resource(url, img_storage, hash_value)
 
     # Create a symbolic link to individual cube storage
-    if os.path.exists(image_cube_file):
-        # Remove existing links
-        os.unlink(image_cube_file)
     os.symlink(img_storage, image_cube_file)
-    local_hash = get_file_sha1(img_storage)
-    return image_cube_file, local_hash
+    return image_cube_file, hash_value
+
 
+def get_cube_additional(
+    url: str,
+    cube_path: str,
+    expected_tarball_hash: str = None,
+    expected_folder_hash: str = None,
+) -> str:
+    """Retrieves additional files of an MLCube. The additional files
+    will be in a compressed tarball file. The function will extract this
+    file and returns the hash of its contents.
 
-def get_cube_additional(url: str, cube_path: str) -> str:
-    """Retrieves and stores the additional_files.tar.gz file from the server
+    Note: there is no scenario of having `expected_tarball_hash == None` and
+    `expected_folder_hash != None` at the same time, since passing the
+    expected_folder_hash is something controlled by the client not the user.
+    If the client had `expected_tarball_hash == None`, then there is no way the
+    client would have expected_folder_hash != None.
 
     Args:
         url (str): URL where the additional_files.tar.gz file can be downloaded.
         cube_path (str): Cube location.
+        expected_tarball_hash (str, optional): expected sha1 hash of tarball file
+        expected_folder_hash (str, optional): expected sha1 hash of uncompressed
+        version of the tarball file
 
     Returns:
-        str: Location where the additional_files.tar.gz file is stored locally.
+        tarball_hash (str): The hash of the downloaded file
+        folder_hash (str): The hash of the uncompressed version of the downloaded file
     """
-    add_path = config.additional_path
-    tball_file = config.tarball_filename
-    return get_cube_file(url, cube_path, add_path, tball_file)
+    additional_files_folder = os.path.join(cube_path, config.additional_path)
 
+    # If the additional_files folder exists, check if it has the expected hash
+    if expected_folder_hash and os.path.exists(additional_files_folder):
+        folder_hash = get_folder_sha1(additional_files_folder)
+        if folder_hash == expected_folder_hash:
```

Comment:
Great point!

The only possible scenario of "a file exists but its hash is invalid" in mind now is when a user manually manipulates the medperf storage. A scenario we shouldn't account for (since whoever can do that can also manipulate the code)

So you are right, I will remove this part of the logic (it is everywhere :) )

Also, I am thinking:

The client will never put a file in its intended destination before the hash is checked. The client will only move the file from the temporary storage to the intended destination **after** the hash has been verified. This means that we don't have to do hash checking again to ensure existing files are valid, right? So I think we can also remove the `ugly` additional_files_folder hash; we can just check that the folder exists.

Am I missing something?